lib/active_model/serializer/adapter/json_api.rb

```diff
@@ -5,60 +5,35 @@ class JsonApi < Base
         extend ActiveSupport::Autoload
         autoload :PaginationLinks
         autoload :FragmentCache
-        autoload :Link
-
-        # TODO: if we like this abstraction and other API objects to it,
-        # then extract to its own file and require it.
-        module ApiObjects
-          module JsonApi
-            ActiveModel::Serializer.config.jsonapi_version = '1.0'
-            ActiveModel::Serializer.config.jsonapi_toplevel_meta = {}
-            # Make JSON API top-level jsonapi member opt-in
-            # ref: http://jsonapi.org/format/#document-top-level
-            ActiveModel::Serializer.config.jsonapi_include_toplevel_object = false
-
-            module_function
-
-            def add!(hash)
-              hash.merge!(object) if include_object?
-            end
-
-            def include_object?
-              ActiveModel::Serializer.config.jsonapi_include_toplevel_object
-            end
-
-            # TODO: see if we can cache this
-            def object
-              object = {
-                jsonapi: {
-                  version: ActiveModel::Serializer.config.jsonapi_version,
-                  meta: ActiveModel::Serializer.config.jsonapi_toplevel_meta
-                }
-              }
-              object[:jsonapi].reject! { |_, v| v.blank? }
-
-              object
-            end
-          end
-        end
+        autoload :ApiObjects
 
         def initialize(serializer, options = {})
           super
           @include_tree = IncludeTree.from_include_args(options[:include])
-          @fieldset = options[:fieldset] || ActiveModel::Serializer::Fieldset.new(options.delete(:fields))
+          @fieldset = options[:fieldset] || ActiveModel::Serializer::Fieldset.new(options[:fields])
         end
 
+        # Build JSON API document.
+        # @return [Hash] document
         def serializable_hash(options = nil)
           options ||= {}
 
-          hash =
-            if serializer.respond_to?(:each)
-              serializable_hash_for_collection(options)
-            else
-              serializable_hash_for_single_resource
-            end
+          primary_data, included = resource_objects_for(serializer)
+
+          is_collection = serializer.respond_to?(:each)
+          hash = {}
+
+          # Unpack data from `primary_data` array when serializing a single resource.
+          hash[:data] = is_collection ? primary_data.map(&:to_h) : primary_data.first.to_h
+
+          hash[:included] = included.map(&:to_h) if included.any?
 
-          ApiObjects::JsonApi.add!(hash)
+          ApiObjects::JsonApiObject.add!(hash)
```

Comment:
Yeah, I suppose that even under the ApiObjects namespace JsonApi doesn't signal it's a json object strongly enough :)